### ISSUE TITLE
ASP.NET Core 2.2 memory leak

### ISSUE THREAD

**Issue by kameleon71** [aspnet/AspNetCore#4564]
Hello all,
I found memory leak when I sent files to API using IFormFile interface.
With version 2.1 everything is ok.

Simple example:
Content-type: multipart/form-data

# controller
```    
[ApiController]
public class FormController : ControllerBase
{
    [HttpPost]
    [Route("[controller]/Insert")]
    public async Task<IActionResult> Insert([FromForm] Form form)
    {
        return Ok();
    }
}
```
# viewmodel
```    
public class Form
{
    public string Name { get; set; }
    public IFormFile File { get; set; }
    public ICollection<DynamicInput> Inputs { get;set; }
}

public class DynamicInput
{
    public string Name { get; set; }
    public IFormFile File { get; set; }
}
```

# process dotnet.exe
![image](https://user-images.githubusercontent.com/45788355/49796161-3ee72180-fd3c-11e8-8128-18d81b9b8ad6.png)

**Comment by davidfowl** [comment]
@kameleon71 Can you provide a full repro? (Client included).

**Comment by kameleon71** [comment]
Here you are: https://github.com/kameleon71/asp.net.core.2.2.memory.leak
This example is based on standard asp.net.core web app in VS2017 (angular client)
Just run, click fetch data and watch memory usage.

**Comment by travisjs** [comment]
@davidfowl any status on this; we have this issue too...

**Comment by travisjs** [comment]
@davidfowl 

This seams to be cause by the following method being called when an array of IFormFiles has to be parsed by the ModelBinder. It does a loop with an upper range of int.MaxValue

`        // Internal for testing.
        internal async Task<CollectionResult> BindComplexCollectionFromIndexes(ModelBindingContext bindingContext, IEnumerable<string> indexNames)
        {
            bool indexNamesIsFinite;
            if (indexNames != null)
            {
                indexNamesIsFinite = true;
            }
            else
            {
                indexNamesIsFinite = false;
                **indexNames = Enumerable.Range(0, int.MaxValue).Select(i => i.ToString(CultureInfo.InvariantCulture));**
            }

            ...
        }`

**Comment by davidfowl** [comment]
cc @pranavkm

**Comment by pranavkm** [comment]
@dougbu is this the same class of issue as https://github.com/aspnet/AspNetCore/issues/7052?

**Comment by Plasma** [comment]
We upgraded to .NET Core 2.2 and memory climbed to 100% over a few days easily. I'm not sure of the cause yet but have taken a few memory dumps but unable to analyze right now, will hopefully understand more soon.

**Comment by dougbu** [comment]
This issue is a duplicate of #4802. The #6793 fix will correct both issues, in 3.0 Preview 3 we hope.

(@pranavkm #7052 relates to binders that succeed unconditionally (and repeatedly) in the same request. The `FormFileModelBinder` isn't in that category.)

**Comment by danyae** [comment]
I have also experienced this problem and for those who doesn't want to revert back to .Net Core 2.1 I suggest using Non-Sequential Indices like described in this article: [https://haacked.com/archive/2008/10/23/model-binding-to-a-list.aspx/](https://haacked.com/archive/2008/10/23/model-binding-to-a-list.aspx/).

In my views I create a string guid index and then use it to bind a model to a list:

```
@{var fileGuid = Guid.NewGuid().ToString();}
<input type="hidden" name="DocumentsForm.TypedFileList.Index" value="@fileGuid" />
<input type="file" name="DocumentsForm.TypedFileList[@fileGuid].File" class="inputfile"  />
```
